File: backend/app/api/languages.py
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from app.core.models import Language

router = APIRouter(prefix="/api", tags=["languages"])

LANGUAGE_NAMES = {
    Language.EN: "English",
    Language.DE: "Deutsch",
    Language.FR: "Français",
    Language.ES: "Español",
    Language.IT: "Italiano",
    Language.JA: "日本語",
    Language.ZH: "中文",
}


class LanguageInfo(BaseModel):
    code: Language
    name: str
    nlp_available: bool
    model: str | None


@router.get("/languages/{code}/demo")
def get_demo_text(request: Request, code: str) -> dict[str, str]:
    """A flawed example text that showcases the language's checks."""
    try:
        language = Language(code)
    except ValueError:
        raise HTTPException(404, "Unknown language")
    path = request.app.state.settings.demos_dir / f"{language.value}.txt"
    if not path.is_file():
        raise HTTPException(404, "No demo text for this language")
    return {"text": path.read_text(encoding="utf-8").strip()}


@router.get("/languages")
def list_languages(request: Request) -> list[LanguageInfo]:
    nlp = request.app.state.nlp
    return [
        LanguageInfo(
            code=language,
            name=LANGUAGE_NAMES[language],
            nlp_available=nlp.is_available(language.value),
            model=nlp.model_name(language.value),
        )
        for language in Language
    ]

File: frontend/src/languages.ts
import type { Messages } from './i18n/messages'
import type { LanguageInfo } from './types'

export function languageLabel(info: LanguageInfo, messages: Messages): string {
  return info.nlp_available ? info.name : messages.basicChecksOnly(info.name)
}

/** Endonym for a language code ("de" → "Deutsch"), the code itself if unknown. */
export function languageName(code: string, languages: LanguageInfo[]): string {
  return languages.find((info) => info.code === code)?.name ?? code
}

// Shown until GET /api/languages responds (or when the backend is unreachable).
export const FALLBACK_LANGUAGES: LanguageInfo[] = [
  { code: 'en', name: 'English', nlp_available: true, model: null },
  { code: 'de', name: 'Deutsch', nlp_available: true, model: null },
  { code: 'fr', name: 'Français', nlp_available: true, model: null },
  { code: 'es', name: 'Español', nlp_available: true, model: null },
  { code: 'it', name: 'Italiano', nlp_available: true, model: null },
  { code: 'ja', name: '日本語', nlp_available: true, model: null },
  { code: 'zh', name: '中文', nlp_available: true, model: null },
]
